File: showroom/user/rtclat/display.c
/*
COPYRIGHT (C) 2000  Paolo Mantegazza (mantegazza@aero.polimi.it)

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Lesser General Public
License as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public
License along with this library; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA.
*/


#include <stdio.h>
#include <stdlib.h>
#include <sys/mman.h>

#include <rtai_mbx.h>
#include <rtai_msg.h>

static RT_TASK *task;

int main(void)
{
	fd_set input;
	struct timeval tv;
	unsigned int msg, ch;
	MBX *mbx;
	long long max = -1000000000, min = 1000000000;
	struct sample { long long min; long long max; int index, ovrn; } samp;

	tv.tv_sec = 0;

 	if (!(task = rt_task_init(nam2num("LATCHK"), 20, 0, 0))) {
		printf("CANNOT INIT MASTER TASK\n");
		exit(1);
	}

 	if (!(mbx = rt_get_adr(nam2num("LATMBX")))) {
		printf("CANNOT FIND MAILBOX\n");
		exit(1);
	}
	mlockall(MCL_CURRENT | MCL_FUTURE);
	rt_make_hard_real_time();

	while (1) {
		rt_mbx_receive(mbx, &samp, sizeof(samp));
		if (max < samp.max) max = samp.max;
		if (min > samp.min) min = samp.min;
		printf("* min: %lld/%lld, max: %lld/%lld average: %d (%d) <Hit [RETURN] to stop> *\n", samp.min, min, samp.max, max, samp.index, samp.ovrn);
		FD_ZERO(&input);
		FD_SET(0, &input);
		tv.tv_usec = 20000;
		if (select(1, &input, NULL, NULL, &tv)) {
			ch = getchar();
			break;
		}
	}

	rt_rpc(rt_get_adr(nam2num("LATCAL")), msg, &msg);
	rt_task_delete(task);
	exit(0);
}

File: showroom/user/python/display.py
#!/usr/bin/python

from rtai_lxrt import *
from rtai_mbx import *
from rtai_msg import *
from rtai_sem import *

TEST_POLL = False

class SAMP(Structure) :
	_fields_ = [("min", c_longlong), 
	            ("max", c_longlong),
		    ("index", c_int),
		    ("ovrn", c_int)]
samp = SAMP(0, 0, 0, 0)

class POLL(Structure) :
	_fields_ = [("max", c_int), 
	            ("min", c_int)]
ufds = POLL(0, 1)

max = -1000000000
min = 1000000000
msg = c_ulong(0)

rt_allow_nonroot_hrt()

task = rt_task_init_schmod(nam2num("LATCHK"), 20, 0, 0, 0, 0xF)
mbx = rt_get_adr(nam2num("LATMBX"))
polld = (rt_poll_s*2)()
rt_make_hard_real_time()

while True :

	if TEST_POLL :
		while True :
			polld[0] = ( mbx, RT_POLL_MBX_RECV )
			polld[1] = ( mbx, RT_POLL_MBX_RECV )
			retval = rt_poll(byref(polld), 2, -200000000)
			libc.printf("POLL: %x %p %p\n", retval, polld[0].what, polld[1].what)
			if polld[0].what == NULL :
				break

	rt_mbx_receive(mbx, byref(samp), sizeof(samp))
	if max < samp.max :
		max = samp.max
	if min > samp.min :
		min = samp.min
	print "* min:", samp.min, "/", min, "max:", samp.max, "/", max, "average: ", samp.index, "(", samp.ovrn, ") <Hit [RETURN] to stop> *\n"
	if libc.poll(byref(ufds), 1, 1) :
		ch = libc.getchar()
		break

print "* SENDING END MESSAGE TO LATENCY *"
rt_rpc_timed(rt_get_adr(nam2num("LATCAL")), msg, byref(msg), nano2count(1000000000))
rt_make_soft_real_time
rt_task_delete(task)
print "* EXITING DISPLAY *"
